Add parameter validation for Cipher module

k-stachowiak opened the pull request:
Based on #1643, this PR adds parameter validation for the Cipher module.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -336,11 +336,12 @@ const int *mbedtls_cipher_list( void );
  *                      structure associated with the given cipher name.
  *
- * \param cipher_name   Name of the cipher to search for.
+ * \param cipher_name   Name of the cipher to search for. This can be \c NULL.
```
I think we should require this to be non-NULL and make the existing check conditional.

k-stachowiak replied:
@hanno-arm I considered doing so, but I was afraid that there is software that depends on the `NULL` return. Also, this is neither `void` nor error code returning function, thus we don't have a common parameter validation scheme for this case. I could add a non-returning optional check, but then it still needs to return `NULL` upon receiving a `NULL` argument.

Do you have a specific idea, how to handle this case? I still think it may be a feasible option to just document the current behavior, but I am willing to handle it in another way.

hanno-arm replied:
Good point on the return value! In that case, I'd suggest we keep the unconditional check, but still document that cipher_name must not be NULL.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -368,21 +369,26 @@ const mbedtls_cipher_info_t *mbedtls_cipher_info_from_type( const mbedtls_cipher
  *                      responsibility of the caller.
+ *
+ * \param ctx           The context to be freed. If this is \c NULL, the
+ *                      function has no effect.
```
Add: If it is not \c NULL, it must point to an initialized cipher context.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,17 +411,19 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
+static inline unsigned int mbedtls_cipher_get_block_size(
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
```
This should be part of parameter validation.

k-stachowiak replied:
There is a technical issue regarding the `static` `inline` functions in this module. We agreed to only define the validating macros in the `.c` files. Thus for these tests to work, the definition would have to be moved to the header file. Do you think this is acceptable in this case? I don't think I see a particular issue with doing so, apart of introducing a new symbol in the public API.

hanno-arm replied:
Oh, good point again. If we move the declaration, it must use the MBEDTLS namespace, and we would need to indicate through a comment like in `platform_util.h` that the macro is internal. Come to think of it, though, maybe it's easier to directly use `MBEDTLS_INTERNAL_VALIDATE[_RET]` here. Ping @mpg

mpg replied:
I agree that the best way is probably to use `MBEDTLS_INTERNAL_VALIDE[_RET]` here (and for inline functions in general).

k-stachowiak replied:
Thanks @mpg and @hanno-arm! I will do so.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,17 +411,19 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
+ * \param ctx    The context of the cipher. This must be initialized.
  *
  * \return       The size of the blocks of the cipher.
  * \return       0 if \p ctx has not been initialized.
```
Minor: Use \c

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,17 +411,19 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
- * \param ctx    The context of the cipher. Must be initialized.
+ * \param ctx    The context of the cipher. This must be initialized.
  *
  * \return       The size of the blocks of the cipher.
```
Minor: The block size of the underlying cipher.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,17 +411,19 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
+static inline unsigned int mbedtls_cipher_get_block_size(
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
```
This should be done through parameter validation.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -427,12 +435,13 @@ static inline unsigned int mbedtls_cipher_get_block_size( const mbedtls_cipher_c
+static inline mbedtls_cipher_mode_t mbedtls_cipher_get_cipher_mode(
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
```
This should be done through parameter validation.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -464,12 +474,13 @@ static inline int mbedtls_cipher_get_iv_size( const mbedtls_cipher_context_t *ct
+static inline mbedtls_cipher_type_t mbedtls_cipher_get_type(
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
```
This should be done through parameter validation.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -444,13 +453,14 @@ static inline mbedtls_cipher_mode_t mbedtls_cipher_get_cipher_mode( const mbedtl
+static inline int mbedtls_cipher_get_iv_size(
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
```
This should be done through parameter validation.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -481,12 +492,13 @@ static inline mbedtls_cipher_type_t mbedtls_cipher_get_type( const mbedtls_ciphe
+static inline const char *mbedtls_cipher_get_name(
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
```
This should be done through parameter validation.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -481,12 +492,13 @@ static inline mbedtls_cipher_type_t mbedtls_cipher_get_type( const mbedtls_ciphe
+    const mbedtls_cipher_context_t *ctx )
 {
     if( NULL == ctx || NULL == ctx->cipher_info )
         return 0;
```
While at it, please add brackets around `0` here and in the other places.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -530,11 +544,12 @@ static inline mbedtls_operation_t mbedtls_cipher_get_operation( const mbedtls_ci
- * \param key           The key to use.
- * \param key_bitlen    The key length to use, in bits.
+ * \param ctx           The generic cipher context. This must be initialized
+ *                      using mbedtls_cipher_info_from_type() or
```
This comment doesn't fit, because `mbedtls_cipher_info_from_type()` returns a cipher information structure, but cannot be used to initialized a cipher context.

hanno-arm replied:
Suggestion: This must be initialized and bound to a cipher information structure.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -553,7 +570,7 @@ int mbedtls_cipher_setkey( mbedtls_cipher_context_t *ctx, const unsigned char *k
  *                      The default passing mode is PKCS7 padding.
  *
- * \param ctx           The generic cipher context.
+ * \param ctx           The generic cipher context. This must be initialized.
```
Suggestion: This must be initialized and bound to a cipher information structure.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -572,8 +590,9 @@ int mbedtls_cipher_set_padding_mode( mbedtls_cipher_context_t *ctx, mbedtls_ciph
  *
- * \param ctx       The generic cipher context.
- * \param iv        The IV to use, or NONCE_COUNTER for CTR-mode ciphers.
+ * \param ctx       The generic cipher context. This must be initialized.
```
Suggestion: This must be initialized and bound to a cipher information structure.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -599,10 +619,12 @@ int mbedtls_cipher_reset( mbedtls_cipher_context_t *ctx );
+ * \param ctx           The generic cipher context. This must be initialized.
+ * \param ad            The additional data to use. If `ad_len > 0`, then this
+ *                      must be a readable buffer of at least \p ad_len Bytes.
  * \param ad_len        the Length of \p ad.
```
Pre-existing: Correct spelling and be more precise, e.g. "The length of \p ad in Bytes."

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -599,10 +619,12 @@ int mbedtls_cipher_reset( mbedtls_cipher_context_t *ctx );
- * \param ad            The additional data to use.
+ * \param ctx           The generic cipher context. This must be initialized.
+ * \param ad            The additional data to use. If `ad_len > 0`, then this
+ *                      must be a readable buffer of at least \p ad_len Bytes.
```
Minor: Add that it may be \c NULL if \p ad_len is zero.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -627,14 +649,16 @@ int mbedtls_cipher_update_ad( mbedtls_cipher_context_t *ctx,
- *                      least \p ilen + block_size. Must not be the same buffer
- *                      as input.
+ * \param output        The buffer for the output data. This must be able to
+ *                      hold at least \p ilen + block_size. This must not be the
```
Minor: Use backticks around the entire arithmetic expression.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -627,14 +649,16 @@ int mbedtls_cipher_update_ad( mbedtls_cipher_context_t *ctx,
  *
- * \param ctx           The generic cipher context.
- * \param input         The buffer holding the input data.
+ * \param ctx           The generic cipher context. This must be initialized.
```
Suggestion: This must be initialized and bound to a key.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -652,9 +676,11 @@ int mbedtls_cipher_update( mbedtls_cipher_context_t *ctx, const unsigned char *i
  *
- * \param ctx           The generic cipher context.
- * \param output        The buffer to write data to. Needs block_size available.
+ * \param ctx           The generic cipher context. This must be initialized.
```
Suggestion: This must be initialized and bound to a key.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -672,10 +698,11 @@ int mbedtls_cipher_finish( mbedtls_cipher_context_t *ctx,
  *
- * \param ctx           The generic cipher context.
- * \param tag           The buffer to write the tag to.
+ * \param ctx           The generic cipher context. This must be initialized.
```
Suggestion: Slightly repeating what's in the \brief section, "This must be initialized, bound to a key, and have just completed a cipher operation through mbedtls_cipher_finish() the tag for which should be written."

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -672,10 +698,11 @@ int mbedtls_cipher_finish( mbedtls_cipher_context_t *ctx,
- * \param tag           The buffer to write the tag to.
+ * \param ctx           The generic cipher context. This must be initialized.
+ * \param tag           The buffer to write the tag to. This must be a readable
+ *                      boffer of at least \p tag_len Bytes.
```
s/boffer/buffer

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -672,10 +698,11 @@ int mbedtls_cipher_finish( mbedtls_cipher_context_t *ctx,
- * \param ctx           The generic cipher context.
- * \param tag           The buffer to write the tag to.
+ * \param ctx           The generic cipher context. This must be initialized.
+ * \param tag           The buffer to write the tag to. This must be a readable
```
s/readable/writable

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -704,18 +732,22 @@ int mbedtls_cipher_check_tag( mbedtls_cipher_context_t *ctx,
+ * \param ctx           The generic cipher context. This must be initialized.
  * \param iv            The IV to use, or NONCE_COUNTER for CTR-mode ciphers.
+ *                      If `iv_len > 0`, this must be a readable buffer of at
+ *                      least \p Bytes.
```
It should be `\p iv_len`.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -704,18 +732,22 @@ int mbedtls_cipher_check_tag( mbedtls_cipher_context_t *ctx,
- * \param input         The buffer holding the input data.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of at least \p ilen Bytes.
  * \param ilen          The length of the input data.
```
Minor: Add "in Bytes".

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -704,18 +732,22 @@ int mbedtls_cipher_check_tag( mbedtls_cipher_context_t *ctx,
- *                      least \p ilen + block_size. Must not be the same buffer
- *                      as input.
+ * \param output        The buffer for the output data. This must be able to
+ *                      hold at least \p ilen + block_size. This must not be the
```
Minor: Add backticks to get correct formatting for the entire arithmetic expression.

hanno-arm commented on library/cipher.c:
```diff
@@ -150,6 +155,7 @@ const mbedtls_cipher_info_t *mbedtls_cipher_info_from_values( const mbedtls_ciph
 
 void mbedtls_cipher_init( mbedtls_cipher_context_t *ctx )
 {
+    CIPHER_VALIDATE( ctx != NULL );
```
I think the NULL check in `mbedtls_cipher_info_from_string()` should be moved to be part of parameter validation.

hanno-arm commented on library/cipher.c:
```diff
@@ -234,13 +245,14 @@ int mbedtls_cipher_setkey( mbedtls_cipher_context_t *ctx, const unsigned char *k
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( iv_len == 0 || iv != NULL );
 
     if( NULL == iv && iv_len == 0 )
```
This check can be removed.

hanno-arm commented on library/cipher.c:
```diff
@@ -335,12 +348,13 @@ int mbedtls_cipher_update( mbedtls_cipher_context_t *ctx, const unsigned char *i
-    }
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( input != NULL );
```
If `ilen == 0`, we should accept `input == NULL`.

hanno-arm replied:
We should allow `input == NULL && ilen == 0`.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -627,14 +649,16 @@ int mbedtls_cipher_update_ad( mbedtls_cipher_context_t *ctx,
- * \param input         The buffer holding the input data.
+ * \param ctx           The generic cipher context. This must be initialized.
+ * \param input         The buffer holding the input data. This must be a
+ *                      readable buffer of at least \p ilen Bytes.
```
We should mention that `input` may be NULL if `ilen == 0`.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -672,10 +698,11 @@ int mbedtls_cipher_finish( mbedtls_cipher_context_t *ctx,
+ * \param ctx           The generic cipher context. This must be initialized.
+ * \param tag           The buffer to write the tag to. This must be a readable
+ *                      boffer of at least \p tag_len Bytes.
  * \param tag_len       The length of the tag to write.
```
In line with other changes, I think we should accept `tag == NULL` if `taglen == 0`.

hanno-arm commented on library/cipher.c:
```diff
@@ -881,8 +900,9 @@ int mbedtls_cipher_set_padding_mode( mbedtls_cipher_context_t *ctx, mbedtls_ciph
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( tag != NULL );
```
Should we require that only if `tag_len > 0`?

hanno-arm commented on library/cipher.c:
```diff
@@ -1004,6 +1034,15 @@ int mbedtls_cipher_auth_encrypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( iv != NULL );
+    CIPHER_VALIDATE_RET( ad_len == 0 || ad != NULL );
+    CIPHER_VALIDATE_RET( input != NULL );
```
We should allow `ilen == 0 && input == NULL`.

hanno-arm replied:
We should allow `input == NULL && ilen == 0`.

hanno-arm commented on library/cipher.c:
```diff
@@ -1051,6 +1090,15 @@ int mbedtls_cipher_auth_decrypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( iv != NULL );
+    CIPHER_VALIDATE_RET( ad_len == 0 || ad != NULL );
+    CIPHER_VALIDATE_RET( input != NULL );
```
We should allow `input == NULL && ilen == 0`.

hanno-arm commented on library/cipher.c:
```diff
@@ -1051,6 +1090,15 @@ int mbedtls_cipher_auth_decrypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( input != NULL );
+    CIPHER_VALIDATE_RET( output != NULL );
+    CIPHER_VALIDATE_RET( olen != NULL );
+    CIPHER_VALIDATE_RET( tag != NULL );
```
We should allow `tag == NULL` if `tag_len == 0`.

hanno-arm commented on library/cipher.c:
```diff
@@ -976,6 +999,13 @@ int mbedtls_cipher_crypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( iv_len == 0 || iv != NULL );
+    CIPHER_VALIDATE_RET( input != NULL );
```
We should allow `input == NULL && ilen == 0`.

hanno-arm commented on library/cipher.c:
```diff
@@ -913,8 +933,11 @@ int mbedtls_cipher_check_tag( mbedtls_cipher_context_t *ctx,
-        MBEDTLS_DECRYPT != ctx->operation )
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    CIPHER_VALIDATE_RET( ctx->cipher_info != NULL );
+    CIPHER_VALIDATE_RET( tag != NULL );
```
We should allow `tag == NULL && tag_len == 0`.

k-stachowiak commented:
@hanno-arm Thank you for a thorough review. I hope I addressed your comments properly. Could you please re-review?

k-stachowiak commented:
The PR has been rebased against #1643

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,20 +413,22 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
- * \return       The size of the blocks of the cipher.
- * \return       0 if \p ctx has not been initialized.
+ * \return       The block size of the underlying cipher.
+ * \return       \c 0 if \p ctx has not been initialized.
```
Pre-existing: This is technically wrong. If the cipher structure is not initialized, the result is entirely undefined. 

Suggestion: Change to "\c 0 if \p ctx has not been setup."

yanesca commented:
I am slightly confused here. It looks like the documentation mandates a tag of length greater than 0, but the parameter validation seems to allow passing a NULL pointer as the tag. Is this intentional?

Also, allowing empty tag for an AEAD mode does not seem like a very good idea to me. The lower level module catches the error in this case, but it would be simply safer and the parameter validation feature would deliver more value if we would not allow it.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -392,8 +400,8 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
- * \param ctx           The context to initialize. May not be NULL.
- * \param cipher_info   The cipher to use.
+ * \param ctx           The context to initialize. This must be initialized.
+ * \param cipher_info   The cipher to use. This may not be \c NULL.
```
After discussing this with @mpg and @gilles-peskine-arm internally, this assumption should be removed. Instead, it should be documented behaviour that the function returns `#MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA` if you pass `NULL` here.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,20 +413,22 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx != NULL, 0 );
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx->cipher_info != NULL, 0 );
 
     return ctx->cipher_info->block_size;
```
Minor and pre-existing: Brackets missing.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -405,20 +413,22 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx );
-    if( NULL == ctx || NULL == ctx->cipher_info )
-        return 0;
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx != NULL, 0 );
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx->cipher_info != NULL, 0 );
```
As discussed internally with @mpg and @gilles-peskine-arm, the check for `ctx->cipher_info != 0` should be unconditional.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -427,15 +437,16 @@ static inline unsigned int mbedtls_cipher_get_block_size( const mbedtls_cipher_c
-    if( NULL == ctx || NULL == ctx->cipher_info )
-        return MBEDTLS_MODE_NONE;
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx != NULL, MBEDTLS_MODE_NONE );
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx->cipher_info != NULL, MBEDTLS_MODE_NONE );
```
As discussed internally with @mpg and @gilles-peskine-arm, the check for `ctx->cipher_info != 0` should be unconditional.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -427,15 +437,16 @@ static inline unsigned int mbedtls_cipher_get_block_size( const mbedtls_cipher_c
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx != NULL, MBEDTLS_MODE_NONE );
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx->cipher_info != NULL, MBEDTLS_MODE_NONE );
 
     return ctx->cipher_info->mode;
```
Minor and pre-existing: Add brackets.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -444,16 +455,17 @@ static inline mbedtls_cipher_mode_t mbedtls_cipher_get_cipher_mode( const mbedtl
-    if( NULL == ctx || NULL == ctx->cipher_info )
-        return 0;
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx != NULL, 0 );
+    MBEDTLS_INTERNAL_VALIDATE_RET( ctx->cipher_info != NULL, 0 );
```
As discussed internally with @mpg and @gilles-peskine-arm, the check for `ctx->cipher_info != 0` should be unconditional.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -464,15 +476,18 @@ static inline int mbedtls_cipher_get_iv_size( const mbedtls_cipher_context_t *ct
-        return MBEDTLS_CIPHER_NONE;
+    MBEDTLS_INTERNAL_VALIDATE_RET(
+        ctx != NULL, MBEDTLS_CIPHER_NONE );
+    MBEDTLS_INTERNAL_VALIDATE_RET(
```
As discussed internally with @mpg and @gilles-peskine-arm, the check for `ctx->cipher_info != 0` should be unconditional.

hanno-arm commented on include/mbedtls/cipher.h:
```diff
@@ -464,15 +476,18 @@ static inline int mbedtls_cipher_get_iv_size( const mbedtls_cipher_context_t *ct
+    MBEDTLS_INTERNAL_VALIDATE_RET(
+        ctx->cipher_info != NULL, MBEDTLS_CIPHER_NONE );
 
     return ctx->cipher_info->type;
```
Minor and pre-existing: Brackets missing.

hanno-arm commented on library/cipher.c:
```diff
@@ -175,8 +181,9 @@ void mbedtls_cipher_free( mbedtls_cipher_context_t *ctx )
-        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    if( cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing.

hanno-arm commented on library/cipher.c:
```diff
@@ -199,11 +206,17 @@ int mbedtls_cipher_setup( mbedtls_cipher_context_t *ctx, const mbedtls_cipher_in
+    CIPHER_VALIDATE_RET( operation == MBEDTLS_ENCRYPT ||
+                         operation == MBEDTLS_DECRYPT );
+    if( ctx->cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing.

hanno-arm commented on library/cipher.c:
```diff
@@ -234,16 +247,15 @@ int mbedtls_cipher_setkey( mbedtls_cipher_context_t *ctx, const unsigned char *k
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    CIPHER_VALIDATE_RET( iv_len == 0 || iv != NULL );
+    if( ctx->cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing.

hanno-arm commented on library/cipher.c:
```diff
@@ -830,10 +850,14 @@ int mbedtls_cipher_finish( mbedtls_cipher_context_t *ctx,
-        MBEDTLS_MODE_CBC != ctx->cipher_info->mode )
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    if( ctx->cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing. 

Minor: Merge the two branches.

hanno-arm commented on library/cipher.c:
```diff
@@ -913,8 +939,12 @@ int mbedtls_cipher_check_tag( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( ctx != NULL );
+    CIPHER_VALIDATE_RET( tag_len == 0 || tag != NULL );
+    if( ctx->cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing.

hanno-arm commented on library/cipher.c:
```diff
@@ -335,12 +350,14 @@ int mbedtls_cipher_update( mbedtls_cipher_context_t *ctx, const unsigned char *i
+    CIPHER_VALIDATE_RET( output != NULL );
+    CIPHER_VALIDATE_RET( olen != NULL );
+    if( ctx->cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing.

hanno-arm commented on library/cipher.c:
```diff
@@ -976,6 +1006,14 @@ int mbedtls_cipher_crypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( ilen == 0 || input != NULL );
+    CIPHER_VALIDATE_RET( output != NULL );
+    CIPHER_VALIDATE_RET( olen != NULL );
+    if( ctx->cipher_info == NULL )
```
Why did we introduce this check?

k-stachowiak replied:
I actually added the ckech in e768b94e86c4703a578afd071b7f364b69de6ba1 , and I can see there were no checks present at that point. In the recent commit I only made this unconditional, but given the context, I may as well remove it.

hanno-arm replied:
Yes, please remove it, as it seems that the check wasn't there before and isn't needed here as far as I can see.

hanno-arm commented on library/cipher.c:
```diff
@@ -1004,6 +1042,16 @@ int mbedtls_cipher_auth_encrypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( output != NULL );
+    CIPHER_VALIDATE_RET( olen != NULL );
+    CIPHER_VALIDATE_RET( tag_len == 0 || tag != NULL );
+    if( ctx->cipher_info == NULL )
```
Why did we introduce this check?

hanno-arm commented on library/cipher.c:
```diff
@@ -1051,6 +1099,16 @@ int mbedtls_cipher_auth_decrypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( output != NULL );
+    CIPHER_VALIDATE_RET( olen != NULL );
+    CIPHER_VALIDATE_RET( tag_len == 0 || tag != NULL );
+    if( ctx->cipher_info == NULL )
```
Why did we introduce this check?

hanno-arm commented on library/cipher.c:
```diff
@@ -1051,6 +1099,16 @@ int mbedtls_cipher_auth_decrypt( mbedtls_cipher_context_t *ctx,
+    CIPHER_VALIDATE_RET( olen != NULL );
+    CIPHER_VALIDATE_RET( tag_len == 0 || tag != NULL );
+    if( ctx->cipher_info == NULL )
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
```
Brackets missing.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+
+    /* mbedtls_cipher_get_block_size() */
+    TEST_INVALID_PARAM_RET( 0, mbedtls_cipher_get_block_size( NULL ) );
+    TEST_ASSERT( mbedtls_cipher_get_block_size( &invalid_ctx ) == 0 );
```
I think we shouldn't add this test; it documents an undesired behaviour that I think we'd ideally like to remove but refrain from doing so because some people might rely on it, even though it's undocumented.

hanno-arm replied:
Ok, this test was there before. In this case, we can as well keep it.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+    TEST_INVALID_PARAM_RET(
+        MBEDTLS_MODE_NONE,
+        mbedtls_cipher_get_cipher_mode( NULL ) );
+    TEST_ASSERT( mbedtls_cipher_get_cipher_mode( &invalid_ctx ) ==
```
Same here.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+
+    /* mbedtls_cipher_get_iv_size() */
+    TEST_INVALID_PARAM_RET( 0, mbedtls_cipher_get_iv_size( NULL ) );
+    TEST_ASSERT( mbedtls_cipher_get_iv_size( &invalid_ctx ) == 0 );
```
Same here.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+    TEST_INVALID_PARAM_RET(
+        MBEDTLS_CIPHER_NONE,
+        mbedtls_cipher_get_type( NULL ) );
+    TEST_ASSERT(
```
Same here.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+
+    /* mbedtls_cipher_get_name() */
+    TEST_INVALID_PARAM_RET( 0, mbedtls_cipher_get_name( NULL ) );
+    TEST_ASSERT( mbedtls_cipher_get_name( &invalid_ctx ) == 0 );
```
Same here.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+    TEST_INVALID_PARAM_RET(
+        MBEDTLS_KEY_LENGTH_NONE,
+        mbedtls_cipher_get_key_bitlen( NULL ) );
+    TEST_ASSERT( mbedtls_cipher_get_key_bitlen( &invalid_ctx ) ==
```
Same here.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+    TEST_INVALID_PARAM_RET(
+        MBEDTLS_OPERATION_NONE,
+        mbedtls_cipher_get_operation( NULL ) );
+    TEST_ASSERT( mbedtls_cipher_get_operation( &invalid_ctx ) ==
```
Same here.

hanno-arm commented on tests/suites/test_suite_cipher.function:
```diff
@@ -21,73 +21,422 @@ void mbedtls_cipher_list(  )
+                               valid_buffer,
+                               valid_bitlen,
+                               valid_operation ) );
+    TEST_ASSERT(
```
Same here.

hanno-arm replied:
The comment applies to all `TEST_ASSERT` instances below; I stop repeating it each time now.

yanesca commented:
Discussed the question about the tag buffer with @hanno-arm and I am happy with the PR after Hanno's comments are addressed.

k-stachowiak commented:
@yanesca Thanks for an extra review! @mpg @hanno-arm I have rebased the PR and added a split of the two kinds of parameter validation tests. Could you please re-review?

gilles-peskine-arm commented:
CI has a failure in a DTLS test, looks like a timeout.

gilles-peskine-arm commented:
retest

mpg commented:
The travis failure indeed looks harmless. I restarted that Travis job, if it passes it will confirm the failure was transient.

mpg commented:
All CIs now passed.

sbutcher-arm commented:
Reviewed, approved and passing CI, so labelling as 'ready to merge'.

mpg commented:
CI only failed in know-intermittent timing test, so as good as a pass

sbutcher-arm commented:
Passes CI and approved, therefore can keep the 'ready to merge' label.
